lucille-core/src/test/java/com/kmwllc/lucille/core/RunnerManagerTest.java

```diff
@@ -1,56 +1,132 @@
+    assertThrows(RunnerManagerException.class, () -> runnerManager.runWithConfig(runId, configId));
 
     // Wait until the run is over
-    runnerManager.waitForRunCompletion();
+    runnerManager.waitForRunCompletion(runId);
 
     // Ensure lucille is not running and make sure we can now kick off a new run
-    assertFalse(runnerManager.isRunning());
-    assertTrue(runnerManager.runWithConfig(config));
+    assertFalse(runnerManager.isRunning(runId));
+    assertTrue(!details.hasThrowable());
 
     // Wait for all lucille threads to finish before exiting
-    runnerManager.waitForRunCompletion();
+    runnerManager.waitForRunCompletion(runId);
   }
 
   @Test
-  public void testWaitForRunCompletion() throws InterruptedException, ExecutionException {
+  public void testWaitForRunCompletion() throws Exception {
     RunnerManager runnerManager = RunnerManager.getInstance();
     Config config = ConfigFactory.load("RunnerManagerTest/sleep.conf");
+    String configId = runnerManager.createConfig(config);
+    String runId = Runner.generateRunId();
 
     // Ensure lucille is not running first
-    assertFalse(runnerManager.isRunning());
+    assertFalse(runnerManager.isRunning(runId));
+
+    runnerManager.runWithConfig(runId, configId);
+    long startTime = System.nanoTime();
+
+    log.info("before waiting for run to complete {} ========", runnerManager.isRunning(runId));
+    assertTrue(runnerManager.isRunning(runId));
+
+    runnerManager.waitForRunCompletion(runId);
 
-    runnerManager.runWithConfig(config);
+    long endTime = System.nanoTime();
+    long durationMillis = (endTime - startTime) / 1_000_000;
 
-    // Ensure lucille is running, wait for it stop and ensure its stopped
-    assertTrue(runnerManager.isRunning());
-    runnerManager.waitForRunCompletion();
-    assertFalse(runnerManager.isRunning());
+    log.info("after waiting for run to complete {} ========", runnerManager.isRunning(runId));
+    assertFalse(runnerManager.isRunning(runId));
+
+    log.info("Test execution time: {} ms", durationMillis);
+  }
+
+
+  @Test
+  public void testSimultaneousRuns() throws Exception {
+    Config config = ConfigFactory.load("RunnerManagerTest/sleep.conf");
+    RunnerManager runnerManager = RunnerManager.getInstance();
+    String configId = runnerManager.createConfig(config);
+    List<String> runIds = new ArrayList();
+    List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+    for (int i = 0; i < 5; i++) {
+      String runId = "test-run-" + i;
+      CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
+        try {
+          runnerManager.runWithConfig(runId, configId);
+        } catch (RunnerManagerException e) {
+          throw new RuntimeException(e);
+        }
+      });
+      futures.add(future);
+      runIds.add(runId);
+    }
+    
+    // Ensure all futures have started
+    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
```

Comment:
this is a way of blocking until all completeable futures are complete

it is not a good way to simply check that they have started, because it also waits for them to finish

it makes the while loop below completely redundant and unnecessary, but what it does not do is make sure we we stay below the 10 second threshold. You could change the join to get(10, TimeUnit.SECONDS) and let a TimeoutException bubble up and fail the test. However, then you would not be testing the behavior of runnermanager.getRunDetails() and details.isDone() which is an important part of this test

why was this line added?

Reply:
Because your previous test had a race condition... 
the call to start -> runnerManager.runWithConfig(runId, configId);  is asynchronous,
and you immediately spin through runnerManager.getRunDetails(x); to check details when they aren't even available yet.  This caused a tricky null pointer which was confusing to diagnose, because it worked locally in debug where it was slower but failed with an npe getting RunDetails in ci/cd. A small 10ms wait for the test's requests to launch would fix it .. but i thought the wait seemed crude.

Reply:
OK, in my comment i was mixing up the futures started in the test code with the futures started by RunnerManager. resolving...